tsl/src/partialize_finalize.c

```diff
@@ -147,6 +148,79 @@ collation_oid_from_name(char *schema_name, char *collation_name)
 	namel = lappend(namel, makeString(collation_name));
 	return get_collation_oid(namel, false);
 }
+
+#if PG14_GE
+/*
+ * Record serialized partials errata here:
+ * ============================================================================
+ * PostgreSQL versions < 14
+ * F_NUMERIC_AVG_DESERIALIZE serialized partial length = X
+ * F_NUMERIC_DESERIALIZE serialized partial length = X + 10
+ *
+ * PostgreSQL versions >= 14
+ * F_NUMERIC_AVG_DESERIALIZE serialized partial length = X + 16
+ * F_NUMERIC_DESERIALIZE serialized partial length = X + 10 + 16
+ *
+ * For F_NUMERIC_AVG_DESERIALIZE and F_NUMERIC_DESERIALIZE partials it suffices to zero out the
+ * extra bytes of pInfcount and nInfcount to support infinity and -infinity in the numeric type.
+ * For more information see: https://www.postgresql.org/message-id/606717.1591924582@sss.pgh.pa.us
+ *
+ * Non-numeric aggregate functions affected:
+ * 1. var_pop(int8)
+ * 2. var_samp(int8)
+ * 3. variance(int8)
+ * 4. stddev_pop(int8)
+ * 5. stddev_samp(int8)
+ * 6. stddev(int8)
+ * ============================================================================
+ * F_NUMERIC_POLY_DESERIALIZE depends on compiler support for HAVE_INT128. As a result,
+ * recompiling the same versions of PostgreSQL and TimescaleDB with a different compiler on the
+ * same database can lead to corruption.
+ *
+ * Non-numeric aggregate functions affected:
+ * 1. var_pop(int4)
+ * 2. var_pop(int2)
+ * 3. var_samp(int4)
+ * 4. var_samp(int2)
+ * 5. variance(int4)
+ * 6. variance(int2)
+ * 7. stddev_pop(int4)
+ * 8. stddev_pop(int2)
+ * 9. stddev_samp(int4)
+ * 10. stddev_samp(int2)
+ * 11. stddev(int4)
+ * 12. stddev(int2)
+ */
+#define NUMERIC_PARTIAL_MISSING_LENGTH (16)
+
+static bytea *
+zero_fill_bytearray(bytea *serialized_partial, size_t missing_length)
+{
+	size_t original_length = VARSIZE_ANY_EXHDR(serialized_partial);
+	size_t desired_length = original_length + missing_length;
+
+	bytea *new_bytea = repalloc(serialized_partial, desired_length + VARHDRSZ);
+
+	SET_VARSIZE(new_bytea, desired_length + VARHDRSZ);
+	void *end_of_serialized_partial = VARDATA(new_bytea) + original_length;
+	memset(end_of_serialized_partial, 0, missing_length);
+
+	return new_bytea;
+}
+#endif
+
+static bytea *
+sanitize_serialized_partial(Oid deserialfnoid, bytea *serialized_partial)
+{
+#if PG14_GE
+	if ((deserialfnoid == F_NUMERIC_DESERIALIZE) || (deserialfnoid == F_NUMERIC_AVG_DESERIALIZE))
+		/* Always add NUMERIC_PARTIAL_MISSING_LENGTH extra bytes because the length is not fixed */
+		return zero_fill_bytearray(serialized_partial, NUMERIC_PARTIAL_MISSING_LENGTH);
```

Comment:
Will we run this `zero_fill_bytearray` for all `bytea` in PG14 and greater?? I mean we don't want to check if we really need to do it before do it?

Reply:
Not sure it would be possible to do that without taking the same amount of time/effort, extra bytes at the end of the structure shouldn't affect anything about how the ones that are okay function and are needed for pre-14 ones to function. So, I dunno, but it seems like it should be fine? I guess the only question is if PG indexes in from the back of the structure anywhere in their code and therefore we'd be overwriting infinities that do exist...we should probably add a test for that somehow?

Reply:
This function is only called inside `PG_CATCH()` after the deserialization has already thrown an exception. So it only gets called if deserialization has failed.

Also, extra bytes at the end of the structure apparently affect things, because postgres throws an exception during `recv` saying essentially that it read fewer bytes than available, and that's bad or something.

Reply:
Ah, interesting. That makes sense. I think the always in the comment was probably confusing, I think I'd add a code comment at the top of this function clarifying that it's only going to be called if we get an error deserializing the numeric value. If @fabriziomello got confused others might as well

Reply:
Yes when this comment was written it was actually always called. Will fix